### ISSUE TITLE
AltGr + <somekey>  as a hotkey for script

### ISSUE THREAD

**Issue by lnksz** [autokey/autokey#366]
## Classification:

UI/Usability

## Reproducibility:

Always

## Version

AutoKey version: 0.95.9

Used GUI (Qt): 

If the problem is known to be present in more than one version, please list all of those.

Installed via: Pamac

Linux Distribution: 5.4.15-2-MANJARO

## Summary

I seem to face a similar issue like  #272 and just cannot get this working.
I use a classic [Hungarian Layout](http://kbdlayout.info/KBDHU/) and I do use the "Alt Gr" key extensively to get the essential symbols for programming.

What I did with AutoHotKey, and try to redo this with AutoKey, is to type `alt_gr+s` (in layout the useless `đ` letter) and get the "ctrl+left" (jump word) functionality.

I have successfully created a simple script with 
```py
keyboard.send_keys('<ctrl>+<left>')
```
And if I set the hotkeys `<ctrl>+<super>+j` I can achieve to invoke the word jumping.
But it would be great to be able to stick to the same combo as on windows.

I cannot select "AltGr + s" on the GUI as a hotkey, and I tried to manually edit the config json file and set the `hotkey[hotKey]` to `đ` but that does not do the trick.

Any ideas?

## Steps to Reproduce (if applicable)

- Create new script `keyboard.send_keys('<ctrl>+<left>')`
- Try to trigger this with a hotkey "alt_gr+<somekey>"

## Expected Results

- alt_gr+s    should be a valid hotkey setting

## Actual Results

- nothing

**Comment by westerlind** [comment]
This is more or less the most important thing I'm missing, not only in Autokey itself (which I otherwise think is very very good!), but in Desktop Linux as a whole, as I can't find any other software that does this either without having major downsides (like being very slow, which xbindkeys is, at least for me).

So, please add support for the AltGr key as a modifier! :)

By the way, I don't think it necessarily has to be visible in the GUI even, as long as it is at least possible to set in the config files and get it to work that way.

**Comment by josephj11** [comment]
Until this is fixed, could `xmodmap` help in any way?

**Comment by westerlind** [comment]
Yes, xmodmap is my workaround at the moment, to move some keys around a bit so I can get closer to what I really want to do, but it's not possible to get it to work like I really want without being able to use AltGr as a modifier key for shortcuts.

**Comment by Mr-Pepe** [comment]
Will this be fixed? Or is there at least a workaround?

I have caps lock as level 3 switch, which frees up the AltGr key to be used as a modifier for shortcuts.

**Comment by johannesjo** [comment]
I'd also love to see this. AltGr usually has fewer shortcuts to it attached which makes it great for global shortcuts...

**Comment by sebastiansam55** [comment]
This appears to exist in the code for autokey already. If one were to manually edit the config files I would think you would be able to use it no problem.

Editing your .json file to set the modifiers for the hotkey like so:
```
[...]
"modifiers": [
         "<alt_gr>"
],
"hotKey": "<whatever>"
[...]
```

https://github.com/autokey/autokey/blob/1c5e3ce8b45fbea49fd17c2b358211434605a89a/lib/autokey/iomediator/key.py

It probably should still be added to the UI hotkey selection.

If anyone can try this out and have it work I will open a PR adding the Alt_GR to the hotkey dialog selection (for GTK at least...)

**Comment by johannesjo** [comment]
@sebastiansam55 thank you very much for the suggestion. I just checked and unfortunately, it does not work. This is how my  `.arr_left.json` looks like;

```json
{
    "type": "phrase",
    "description": "arr_left",
    "modes": [
        3
    ],
    "usageCount": 18,
    "prompt": false,
    "omitTrigger": false,
    "matchCase": false,
    "showInTrayMenu": false,
    "abbreviation": {
        "abbreviations": [],
        "backspace": true,
        "ignoreCase": false,
        "immediate": false,
        "triggerInside": false,
        "wordChars": "[\\w]"
    },
    "hotkey": {
        "modifiers": [
            "<alt_gr>"
        ],
        "hotKey": "j"
    },
    "filter": {
        "regex": null,
        "isRecursive": false
    },
    "sendMode": "kb"
}
```

**Comment by sebastiansam55** [comment]
Dang. This might be hard to fix without a ISO keyboard (which I don't have access to). Can you run `xev -event keyboard` and mash the ALT_GR button a few times and see what it says?

From reading it seems like ALT_GR is translated to ISO_Level3_Shift? Not sure.

Running `xdotool key ISO_Level3_Shift` and watching the output of autokey verbose shows it is reading a Key.ALT_GR event...

**Comment by johannesjo** [comment]
For my external keyboard:
```
KeyPress event, serial 28, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3060789, (2560,-443), root:(2560,672),
    state 0x0, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3060902, (2560,-443), root:(2560,672),
    state 0x80, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3061277, (2560,-443), root:(2560,672),
    state 0x0, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3061405, (2560,-443), root:(2560,672),
    state 0x80, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```

For my laptop keyboard:
```
KeyRelease event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3113945, (2436,-461), root:(2436,654),
    state 0x80, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3114689, (2436,-461), root:(2436,654),
    state 0x0, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3114793, (2436,-461), root:(2436,654),
    state 0x80, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3115329, (2436,-461), root:(2436,654),
    state 0x0, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3115433, (2436,-461), root:(2436,654),
    state 0x80, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3115817, (2436,-461), root:(2436,654),
    state 0x0, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 29, synthetic NO, window 0x2000001,
    root 0x961, subw 0x0, time 3115929, (2436,-461), root:(2436,654),
    state 0x80, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

MappingNotify event, serial 29, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 248

```

**Comment by sebastiansam55** [comment]
Does the output of both `xev` include `MappingNotify` for both? 

Can you have `autokey-gtk -l` (verbose) running when you hit Alt_GR, does it ungrab and regrab all of the hotkeys?

Before mine starts ungrabbing/regrabbing it says `Recorded keymap change event` you might want to try running autokey in Qt as that version doesn't seem to freak out the  same way

That is what is happening on my end and it is probably related.

**Comment by sebastiansam55** [comment]
Does altgr function like the shift key or the capslock key? 

https://en.wikipedia.org/wiki/AltGr_key

Says it is like the shift key.

**Comment by sebastiansam55** [comment]
This under autokey-qt:
https://github.com/sebastiansam55/autokey/commit/7d892b5b661dd86926a526d94cedf31da8ccbdc6

Works to trigger a script bound to <alt_gr>+1 triggered via `xdotool keydown ISO_Level3_Shift; xdotool key 1; xdotool keyup ISO_Level3_Shift`

**Comment by johannesjo** [comment]
> Can you have `autokey-gtk -l` (verbose) running when you hit Alt_GR, does it ungrab and regrab all of the hotkeys?

``` 
2020-11-20 18:20:33,979 DEBUG - service - Input queue at end of handle_keypress: deque(['﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', '﹠', 'l', 'k', 'j', '﹠', 'Ƴ', 'ࣽ', '﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠'], maxlen=150)
2020-11-20 18:20:34,128 DEBUG - service - Raw key: 'l', modifiers: [], Key: Ƴ
2020-11-20 18:20:34,128 DEBUG - service - Window visible title: 'AltGr + <somekey> as a hotkey for script · Issue #366 · autokey/autokey - Google Chrome', Window class: 'google-chrome.Google-chrome'
2020-11-20 18:20:34,129 DEBUG - service - Input queue at end of handle_keypress: deque(['﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', '﹠', 'l', 'k', 'j', '﹠', 'Ƴ', 'ࣽ', '﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ'], maxlen=150)
2020-11-20 18:20:34,210 DEBUG - iomediator - Key.ALT_GR released
2020-11-20 18:20:34,386 DEBUG - iomediator - Key.ALT_GR pressed
2020-11-20 18:20:34,547 DEBUG - service - Raw key: 'k', modifiers: [], Key: ΢
2020-11-20 18:20:34,547 DEBUG - service - Window visible title: 'AltGr + <somekey> as a hotkey for script · Issue #366 · autokey/autokey - Google Chrome', Window class: 'google-chrome.Google-chrome'
2020-11-20 18:20:34,548 DEBUG - service - Input queue at end of handle_keypress: deque(['﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', '﹠', 'l', 'k', 'j', '﹠', 'Ƴ', 'ࣽ', '﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2'], maxlen=150)
2020-11-20 18:20:34,634 DEBUG - iomediator - Key.ALT_GR released
2020-11-20 18:20:34,707 DEBUG - service - Raw key: 'k', modifiers: [], Key: k
2020-11-20 18:20:34,707 DEBUG - service - Window visible title: 'AltGr + <somekey> as a hotkey for script · Issue #366 · autokey/autokey - Google Chrome', Window class: 'google-chrome.Google-chrome'
2020-11-20 18:20:34,708 DEBUG - service - Input queue at end of handle_keypress: deque(['﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', '﹠', 'l', 'k', 'j', '﹠', 'Ƴ', 'ࣽ', '﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', 'k'], maxlen=150)
2020-11-20 18:20:34,747 DEBUG - iomediator - Key.ALT_GR pressed
2020-11-20 18:20:34,811 DEBUG - service - Raw key: 'j', modifiers: [], Key: ﹠
2020-11-20 18:20:34,811 DEBUG - service - Window visible title: 'AltGr + <somekey> as a hotkey for script · Issue #366 · autokey/autokey - Google Chrome', Window class: 'google-chrome.Google-chrome'
2020-11-20 18:20:34,811 DEBUG - service - Input queue at end of handle_keypress: deque(['﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', '﹠', 'l', 'k', 'j', '﹠', 'Ƴ', 'ࣽ', '﹠', '\u03a2', 'Ƴ', 'ࣽ', '﹠', 'Ƴ', '\u03a2', 'k', '﹠'], maxlen=150)
2020-11-20 18:20:34,874 DEBUG - iomediator - Key.ALT_GR released
2020-11-20 18:20:35,162 DEBUG - iomediator - Key.ALT_GR pressed
2020-11-20 18:20:35,233 DEBUG - iomediator - Key.ALT_GR released
2020-11-20 18:20:35,282 DEBUG - iomediator - Key.ALT_GR pressed
2020-11-20 18:20:35,361 DEBUG - iomediator - Key.ALT_GR released
```

**Comment by lnksz** [comment]
as  #478 resolves this I close the issue

**Comment by katutxakurra** [comment]
I tested the develop branch a043bf9 in the latest KDE Neon and, in addition to this not being implemented for qt, I found some weird behavior. From the GUI I set the first phrase <altgr>+1 and then I tested in the text editor Kate. The terminal shows
2020-12-11 22:33:26,185 INFO - autokey.service - Matched Phrase "First phrase" with hotkey and prompt=False
But the phrase is not printed in Kate. Instead, the cursor moves to the first position, top left, and writes " !on  numph " without the double quotes. This also happens in KWrite but it works as expected in LibreOffice and Eclipse editors.
When I change the hotkey to <alt>+1 or just 1, it works in all text editors.
